src/commcare_cloud/fab/fabfile.py

```diff
@@ -314,6 +314,9 @@ def _setup_release(keep_days=2, full_cluster=True):
     :param keep_days: The number of days to keep this release before it will be purged
     :param full_cluster: If False, only setup on webworkers[0] where the command will be run
     """
+    if not env.password:
```

Comment:
This just ensures the sudo password is populated for any subsequent sudo(...) commands. By explicitly prompting the user already here, no underlying "parallel ambiguity" w.r.t. sudo password inputs should exist.

Reply:
@Charl1996 what sudo password is it expecting here? not entirely comfortable giving any executable code the carte blanche to my _local_ machine's root password (which was apparently the input it was looking for)

Reply:
ah it looks like if i type anything then it still goes through. can this check also see if the command is being run in a monolith context? It really adds another annoying prompt in the workflow for regular SaaS deploys.

Reply:
Another thought is that it doesn't seem great to be storing ANY passwords in a variable, but rather using the ansible `sudo()` commands properly instead. I think I'm struggling to understand why this check is here at all.

Reply:
+1 - this should not be here. If the machine's sudoers file is set up correctly it should not be necessary

Reply:
https://github.com/dimagi/commcare-cloud/pull/4700

Reply:
@biyeun I'm not _completely_ satisfied with this approach myself, but found that it works great in monolith context. I think your suggestion w.r.t. checking monolith context is great! I will tinker a bit to add it. 

**So, why the check at all?**
During the monolith install deploy step, the process would be stopped due to fatal error: 

> Fatal error: Needed to prompt for a connection or sudo password (host: <host>), but input would be ambiguous in parallel mode
 
This was due to the **@parallel** decorator decorating a function which simply creates a directory with **sudo()** (see function _create_code_dir_ in _src/commcare_cloud/fab/operations/release.py_). The real issue seems to be the fact that it wants to prompt a sudo password because it does not know it at this point. Apparently when using the **sudo()** command you can store the password in the fabric env so the command won't prompt again, that's why I reasoned to prompt the user first and get the password, so by the time the **sudo()** command executes it already knows the password.

> "ah, it looks like if I type anything then it still goes through" 

This is concerning. When I tested it with a wrong password it failed at the sudo command step creating the directory. It seems there's factors in play in a non-monolith setup which I don't know of. Correct me if I'm wrong, but it doesn't look like this new logic is affecting a non-monolith context apart from the actual prompting...?

Reply:
@snopoke I had a look at the sudoers file and everything looks in order, so I didn't think it could be that?

**What I looked for**
I basically made sure my user is listed under

> User privilege specification

with 

> ALL=(ALL:ALL) ALL

I might be completely wrong on this as well, this is kind of fresh ground for me...should I be looking for something else?

Reply:
The deploy generally uses the `ansible` user unless otherwise specified. For deploy the ansible user needs sudo permissions to perform actions as the `cchq` user.

If you know that command that was requesting a password you could test it out on the machine:
```
cchq <env> ssh ansible@<machine name>

ansible@X~$ sudo -u cchq <command>
```

Not all commands are allowed so it may be that the command being attempted is not in the sudoers file correctly.

Since this isn't an issue on any of our existing envs I'm sure it has something to do with how the test env is being set.